File: packages/schemas/ts/src/event.ts
import { z } from "zod";
import { SchemaVersionV1 } from "./schema-version";

/**
 * Domain research events (PRD §6 "Event"): ordered, append-only, reconnectable
 * projections of research activity. CLOSED envelope (unknown fields are
 * rejected in every stack) with a permissive, already-redacted payload.
 * Payloads must never contain secrets or raw provider prompts/responses.
 */
export const EventType = z.enum([
  "task.created",
  "task.started",
  "task.progress",
  "task.checkpoint",
  "task.completed",
  "task.failed",
  "task.skipped",
  "run.started",
  "run.paused",
  "run.completed",
  "run.failed",
  "run.cancelled",
  "run.incremental_report",
  "plan.drafted",
  "plan.approved",
  "plan.rejected",
  "plan.superseded",
  "source.discovered",
  "source.fetched",
  "source.evaluated",
  "claim.created",
  "claim.updated",
  "claim.superseded",
  "review.requested",
  "review.resolved",
]);
export type EventType = z.infer<typeof EventType>;

export const EventSchema = z
  .object({
    schema_version: SchemaVersionV1,
    event_id: z.string().min(1),
    sequence: z.number().int().min(1),
    occurred_at: z.string(),
    run_id: z.string().min(1).nullable().optional(),
    task_id: z.string().min(1).optional(),
    project_id: z.string().min(1),
    type: EventType,
    summary: z.string().optional(),
    payload: z.record(z.unknown()),
  })
  .strict();
export type Event = z.infer<typeof EventSchema>;

File: packages/schemas/rust/src/event.rs
//! Domain research event binding (PRD §6 "Event").
//!
//! Ordered, append-only, reconnectable projections of research activity.
//! Closed envelope (deny_unknown_fields) with a permissive, already-redacted
//! payload; payloads must never contain secrets or raw provider responses.

use serde::{Deserialize, Serialize};

use crate::schema_version::SchemaVersionV1;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum EventType {
    #[serde(rename = "task.created")]
    TaskCreated,
    #[serde(rename = "task.started")]
    TaskStarted,
    #[serde(rename = "task.progress")]
    TaskProgress,
    #[serde(rename = "task.checkpoint")]
    TaskCheckpoint,
    #[serde(rename = "task.completed")]
    TaskCompleted,
    #[serde(rename = "task.failed")]
    TaskFailed,
    #[serde(rename = "task.skipped")]
    TaskSkipped,
    #[serde(rename = "run.started")]
    RunStarted,
    #[serde(rename = "run.paused")]
    RunPaused,
    #[serde(rename = "run.completed")]
    RunCompleted,
    #[serde(rename = "run.failed")]
    RunFailed,
    #[serde(rename = "run.cancelled")]
    RunCancelled,
    #[serde(rename = "plan.drafted")]
    PlanDrafted,
    #[serde(rename = "plan.approved")]
    PlanApproved,
    #[serde(rename = "plan.rejected")]
    PlanRejected,
    #[serde(rename = "plan.superseded")]
    PlanSuperseded,
    #[serde(rename = "source.discovered")]
    SourceDiscovered,
    #[serde(rename = "source.fetched")]
    SourceFetched,
    #[serde(rename = "source.evaluated")]
    SourceEvaluated,
    #[serde(rename = "claim.created")]
    ClaimCreated,
    #[serde(rename = "claim.updated")]
    ClaimUpdated,
    #[serde(rename = "claim.superseded")]
    ClaimSuperseded,
    #[serde(rename = "review.requested")]
    ReviewRequested,
    #[serde(rename = "review.resolved")]
    ReviewResolved,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Event {
    pub schema_version: SchemaVersionV1,
    pub event_id: String,
    pub sequence: i64,
    pub occurred_at: String,
    pub run_id: Option<String>,
    pub task_id: Option<String>,
    pub project_id: String,
    #[serde(rename = "type")]
    pub event_type: EventType,
    pub summary: Option<String>,
    pub payload: serde_json::Value,
}

use crate::ContractValid;

impl ContractValid for Event {
    fn check_contract(&self) -> Result<(), String> {
        if self.event_id.is_empty() || self.project_id.is_empty() {
            return Err("event_id and project_id must not be empty".into());
        }
        if self.sequence < 1 {
            return Err("sequence must be at least 1".into());
        }
        if let Some(run_id) = &self.run_id {
            if run_id.is_empty() {
                return Err("run_id must not be empty when present".into());
            }
        }
        if let Some(task_id) = &self.task_id {
            if task_id.is_empty() {
                return Err("task_id must not be empty when present".into());
            }
        }
        Ok(())
    }
}
